scylla-cdc/src/log_reader.rs

```diff
@@ -141,10 +156,9 @@ impl CDCReaderWorker {
                             })
                         })
                         .collect();
-                } else if had_first_generation {
-                    // FIXME: There may be another generation coming in the future with timestamp < end_timestamp
-                    // that could have been missed because of earlier fetching failures.
-                    // More on this here https://github.com/piodul/scylla-cdc-rust/pull/10#discussion_r826865162
```

Comment:
If we are here, then it means that all stream readers read up to the `end_timestamp` and stopped. The problem that may occur here is that a generation `G2` which is less than the `end_timestamp` appeared in scylla but the generation fetcher was delayed for some reason and we don't know about it yet. We need to check if `G2` exists - if it does, we need to start reading from it, otherwise we can stop.

The idea is to query the generations table here and see if such `G2` exists or not.

It looks like your solution attempts to solve this problem by waiting some time on the `generation_receiver` channel and it assumes that `G2` exists if it appears on the channel within some timeout. You cannot assume that if something doesn't happen within some timeout then it won't happen at all - in an asynchronous, non-real-time system there are many factors which may delay execution of code, e.g.:

- The runtime scheduler may have a very long queue of tasks and your code is at the end of it,
- The OS thread the application executes in may be preempted by other processes,
- A request to scylla takes long because the cluster is overloaded,

Etc.

----

The simplest way I can think of to fix the problem without using timeouts like that is:

1. Use `fetch_next_generation` here to see if there is a new generation which is lower than `end_timestamp`.
2. If there is, then continue the main loop. The generation fetcher will eventually notice this generation and will return it on the `generation_receiver` channel, then the log reader will read the range `G2..end_timestamp`.
3. If not, then quit.